feat: update mcp-core prompt to match spec, add mcp list/get to mcp-client

kalvinnchau opened the pull request:
Update `Prompt` structs for align with MCP spec, and add list/get prompts to the `mcp-client`

## Motivation and Context
* [Prompts](https://modelcontextprotocol.io/docs/concepts/prompts#prompt-structure) `description` and `arguments` are optional fields, use optional in the structs
   * similarly `PromptArguments` `description` and `required` are optional too
* add `list_prompts` and `get_prompt` method to the `McpClientTrait`
* add implementation of `list_prompts` and `get_prompt` to call `prompts/list` and `prompts/get` respectively
* update `stdio` and `sse` clients to handle `JsonRpcMessage::Error` responses to send back to the user instead of doing nothing

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->
Tested locally with goose, and tested listing, and getting prompts and getting errors
Added implementation to `crates/mcp-server/src/main.rs` and usage of it in `crates/mcp-client/examples/stdio_integration.rs`

## Breaking Changes
<!-- Will users need to update their code or configurations? -->
* The `Prompt` struct API has changed `description` and `arguments` to be `Option<T>`
* The `PromptArgument` struct API has changed `description` and `required` to be `Option<T>`
* `Router`: `get_prompt` and `list_prompt` are now required to be implemented by types that want to `impl Router` to keep it consistent with how `tools` and `resources` are handled.


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

salman1993 commented:
nit: you can add a prompt in the mcp-server example here: https://github.com/modelcontextprotocol/rust-sdk/blob/main/crates/mcp-server/src/main.rs. that'll make it easier to test from this mcp-client example: https://github.com/modelcontextprotocol/rust-sdk/blob/main/crates/mcp-client/examples/stdio_integration.rs

kalvinnchau commented:
Added the examples, and changed the Router methods to be in-line with tools/resources, which is a breaking change since implementing types will need to define those.

```
❯ cargo run -p mcp-client --example stdio_integration
...
Prompts: ListPromptsResult { prompts: [Prompt { name: "example_prompt", description: Some("This is an example prompt that takes one required agrument, message"), arguments: Some([PromptArgument { name: "message", description: Some("A message to put in the prompt"), required: Some(true) }]) }] }

...

Prompt: GetPromptResult { description: Some("This is an example prompt with your message here: 'hello there!'"), messages: [PromptMessage { role: User, content: Text { text: "This is an example prompt with your message here: 'hello there!'" } }] }
```

salman1993 commented:
lgtm! i don't have write access in this repo
